From 2627bc2840aa43caa6564f3bd930e1083c92e95b Mon Sep 17 00:00:00 2001
From: Manjunath <manjunathx.b.s@intel.com>
Date: Fri, 7 Sep 2012 18:52:14 +0530
Subject: [PATCH 42/56] Playing OGG format throws error that playback not
 supported.

BZ: 49659

This patch contains Changes in OggExtractor.cpp. The
"timeUs" is calculated only when mGranulePosition > 0.

Orig-Change-Id: Ied8556f1217d24505a341fe52e1ff326bcbb0ee6
Category: aosp improvement
Domain: AUDIO_MEDIA-Audio Codec
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: Ice7a8ebc8a1f86d8ee98112e15908156a55c1f1d
Signed-off-by: Manjunath <manjunathx.b.s@intel.com>
Reviewed-on: http://android.intel.com:8080/77918
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Reviewed-by: Kandasamy, Muthukumar <muthukumar.kandasamy@intel.com>
Reviewed-by: M, Arulselvan <arulselvan.m@intel.com>
Tested-by: M, Arulselvan <arulselvan.m@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
Signed-off-by: Arulselvan M <arulselvan.m@intel.com>
---
 media/libstagefright/OggExtractor.cpp | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/media/libstagefright/OggExtractor.cpp b/media/libstagefright/OggExtractor.cpp
index b763dd9..610b700 100644
--- a/media/libstagefright/OggExtractor.cpp
+++ b/media/libstagefright/OggExtractor.cpp
@@ -457,7 +457,7 @@ status_t MyVorbisExtractor::readNextPacket(MediaBuffer **out) {
     *out = NULL;
 
     MediaBuffer *buffer = NULL;
-    int64_t timeUs = -1;
+    int64_t timeUs = -1ll;
 
     for (;;) {
         size_t i;
@@ -509,7 +509,7 @@ status_t MyVorbisExtractor::readNextPacket(MediaBuffer **out) {
                 // this packet, we also stamp every packet in this page
                 // with the same time. This needs fixing later.
 
-                if (mVi.rate) {
+                if (mVi.rate && ((int64_t)mCurrentPage.mGranulePosition >= 0)) {
                     // Rate may not have been initialized yet if we're currently
                     // reading the configuration packets...
                     // Fortunately, the timestamp doesn't matter for those.
@@ -519,6 +519,16 @@ status_t MyVorbisExtractor::readNextPacket(MediaBuffer **out) {
             }
             buffer = tmp;
 
+            if (timeUs > 0) {
+                int64_t durationUs = 0;
+                if (mMeta->findInt64(kKeyDuration, &durationUs)) {
+                    if (durationUs < timeUs) {
+                        ALOGV("Reached End of File");
+                        return ERROR_END_OF_STREAM;
+                    }
+                }
+            }
+
             ssize_t n = mSource->readAt(
                     dataOffset,
                     (uint8_t *)buffer->data() + buffer->range_length(),
-- 
2.7.4

